### PR TITLE
Re-use identical tile's geometry and oriented the main normal on the Z axis

### LINKED ISSUE #558: Mesh from features should be in local coordinate
Mesh from features should be in local coordinate

## Context
Suggestion of an animation that scale buildings on Z axis
https://github.com/iTowns/itowns/pull/519#pullrequestreview-72184139

The animation works in the WFS planar example : We can see that the builldings is growing.
But it doesn't work in WFS globe example.

It's because feature is in global coordinates, and not in local coordinates.

## Possible Solution
Suggestion of a way to do it 
https://github.com/iTowns/itowns/pull/519#discussion_r147365812

### PR DESCRIPTION
**Pull request by gchoqueux**
## Description
Re-use identical tile's geometry and oriented the main normal on the Z axis
* The normal in the center of the geometry is aligned with the Z axis.
* Bounding boxes can use the Parent's worldMatrix (tile).

The geometry normal is computed with correct geodesic normal.
The bounding boxes is calculated by THREE.js.

## Motivation and Context
* reduce memory
* increase performance
* work tile's space, by example wfs (fix #558 issue)
* simplify obb computation
* reduce/fix obb calculation errors see #87 issues
* fix obb/culling for pano tile

**WARNING**
change subdivision count in PLANAR test, the bounding sphere are bigger, so the subdivision are changed

### CODE REVIEW

**Comment by peppsac** [comment]
3 fixes, 2 feat, 1 refactor => please, split this PR.

**Comment by gchoqueux** [comment]
@peppsac 
normally, these commits should be squashed, because  last 2 commits fix the first: 
* refactor (core): use common geometry in meshTile b5b20f2 
* fix (test): !change subdivision count in PLANAR test! 11d9e71
* fix (protocols): wfs, adapt the mesh transformation with the tile space e2e7c8b

and the 2 debugging features are needed to verify how this PR works :
* feat (examples): add gui debug in planar in debug mode only 70a6cb0 
* feat (core) : add bounding sphere helper in debug mode 65220e3

I remove 
fix (core): improve search for min/max altitude in xbil buffer  663b96d5

File: index.html [hunk 1]
```diff
@@ -45,9 +45,9 @@
 
             var promises = []
             promises.push(itowns.Fetcher.json('examples/layers/JSONLayers/Ortho.json').then(function(result) { return globeView.addLayer(result); }));
-            promises.push(itowns.Fetcher.json('examples/layers/JSONLayers/OrthosCRS.json').then(function(result) { return globeView.addLayer(result); }));
-            promises.push(itowns.Fetcher.json('examples/layers/JSONLayers/ScanEX.json').then(function(result) { return globeView.addLayer(result); }));
-            promises.push(itowns.Fetcher.json('examples/layers/JSONLayers/Region.json').then(function(result) { return globeView.addLayer(result); }));
+            // promises.push(itowns.Fetcher.json('examples/layers/JSONLayers/OrthosCRS.json').then(function(result) { return globeView.addLayer(result); }));
```
Comment (peppsac) [inline index.html]: You forgot to remove this temporary change

File: src/Core/Geographic/Coordinates.js [hunk 2]
```diff
@@ -136,7 +136,8 @@ function _convert(coordsIn, newCrs, target) {
         }
     } else {
         if (coordsIn.crs === 'EPSG:4326' && newCrs === 'EPSG:4978') {
-            const cartesian = ellipsoid.cartographicToCartesian(coordsIn);
+            target._normal = ellipsoid.geodeticSurfaceNormalCartographic(coordsIn);
```
Comment (peppsac) [inline src/Core/Geographic/Coordinates.js]: Doing:
```js
target._normal = coordsIn.geodesicNormal;
```
Would avoid recomputing the normal if it was already done for `coordsIn`

File: src/Core/Math/Ellipsoid.js [hunk 3]
```diff
@@ -56,10 +56,10 @@ Ellipsoid.prototype.setSize = function setSize(size) {
 };
 
 
-Ellipsoid.prototype.cartographicToCartesian = function cartographicToCartesian(coordCarto) {
+Ellipsoid.prototype.cartographicToCartesian = function cartographicToCartesian(coordCarto, normal) {
```
Comment (peppsac) [inline src/Core/Math/Ellipsoid.js]: You must change the code to not modify the `normal` given by the user (modifying parameters makes the API error-prone)

File: src/Core/Prefab/GlobeView.js [hunk 4]
```diff
@@ -163,6 +164,31 @@ export function createGlobeLayer(id, options) {
         return false;
     }
 
+    // return common extent to pool the geometries
+    // the geometry in common extent is identical to the existing input
+    // with a transformation (translation, rotation)
+    wgs84TileLayer.getCommonGeometryExtent = (extent) => {
```
Comment (peppsac) [inline src/Core/Prefab/GlobeView.js]: These functions don't belong to `GlobeView` because:
- the panorama layers will want to use them
- they are tile geometries implementation details

Comment (peppsac) [reply src/Core/Prefab/GlobeView.js]: Also the naming is wrong: you're basically doing a modulo operation on the longitude component (why `sizeLatitude`?) of the extent so the name has to express this.
Btw, why aren't you doing the same thing for the other angle? (I guess it has to do with the ellipsoid shape, but you may want to add a comment in the code)

File: src/Core/Prefab/GlobeView.js [hunk 5]
```diff
@@ -163,6 +164,31 @@ export function createGlobeLayer(id, options) {
...
+    };
+
+    const axisZ = new THREE.Vector3(0, 0, 1);
+    const axisY = new THREE.Vector3(0, 1, 0);
+    const quatToAlignLongitude = new THREE.Quaternion();
+    const quatToAlignLatitude = new THREE.Quaternion();
+
+    // return rotation to transform tile to position it on ellipsoid
+    // this transformation take into account the transformation of the parents
+    wgs84TileLayer.getQuaternionTile = (tile) => {
```
Comment (peppsac) [inline src/Core/Prefab/GlobeView.js]: Wrong naming: the name doesn't express what the function does.
And the argument should be an extent, since you don't use any other properties of the `tile`.

File: src/Core/Prefab/Planar/PlanarTileBuilder.js [hunk 6]
```diff
@@ -50,19 +50,8 @@ PlanarTileBuilder.prototype.vProjecte = function vProjecte(v, params)
...
-        params.extent.north(),
-        0).sub(center);
-    const min = new THREE.Vector3(
-        params.extent.west(),
-        params.extent.south(),
-        0).sub(center);
-    const translate = new THREE.Vector3(0, 0, 0);
-    // normal is up vector
-    return new OBB(min, max, undefined, translate);
+PlanarTileBuilder.prototype.OBB = function OBBFn(params, boundingBox) {
```
Comment (peppsac) [inline src/Core/Prefab/Planar/PlanarTileBuilder.js]: Why keep the `params` parameter if it's unused?

Comment (peppsac) [reply src/Core/Prefab/Planar/PlanarTileBuilder.js]: (same question for `BuilderEllipsoidTile`)

File: src/Core/Scheduler/Providers/TileProvider.js [hunk 7]
```diff
@@ -49,50 +41,84 @@ TileProvider.prototype.preprocessDataLayer = function preprocessLayer(layer, vie
...
+    if (!this.cacheGeometry[level]) {
+        this.cacheGeometry[level] = new Map();
+    }
 
-    var tile = new TileMesh(geometry, params);
+    const ce = command.layer.getCommonGeometryExtent(extent);
+    const south = ce.south().toFixed(8);
+
+    // build geometry if doesn't exist
+    if (!this.cacheGeometry[level].get(south)) {
```
Comment (peppsac) [inline src/Core/Scheduler/Providers/TileProvider.js]: This is wrong: the `south` value alone can't serve as a cache key.
You must at least depend on `command.layer.disableSkirt` (and probably on ellipsoidSize - even if currently this cannot be changed) and the `builder` used.

File: src/Core/Scheduler/Providers/TileProvider.js [hunk 8]
```diff
@@ -49,50 +41,84 @@ TileProvider.prototype.preprocessDataLayer = function preprocessLayer(layer, vie
...
+        this.cacheGeometry[level].set(south, defer());
```
Comment (peppsac) [inline src/Core/Scheduler/Providers/TileProvider.js]: Why is it necessary to use a Promise mechanism here (instead of storing the geometry directly in the cache)?

File: src/Core/TileMesh.js [hunk 9]
```diff
@@ -25,11 +25,10 @@ function TileMesh(geometry, params) {
     this.extent = params.extent;
 
     this.geometry = geometry;
-    this.normal = params.center.clone().normalize();
 
-    this.boundingSphereOffset = new THREE.Vector3();
+    this.obb = this.geometry.OBB.clone();
```
Comment (peppsac) [inline src/Core/TileMesh.js:29]: What's the point of `this.obb`?

File: src/Core/TileMesh.js [hunk 10]
```diff
@@ -43,14 +42,17 @@ function TileMesh(geometry, params) {
     this.layerUpdateState = {};
 
     this.material.setUuid(this.id);
+
+    this.invWorldQuaternion = new THREE.Quaternion();
```
Comment (peppsac) [inline src/Core/TileMesh.js]: Why do you need `invWorldQuaternion`?

File: src/Renderer/ThreeExtended/Feature2Mesh.js [hunk 11]
```diff
@@ -219,13 +221,14 @@ function coordinateToPolygonExtruded(coordinates, properties, options) {
...
     let nbVertices = 0;
-
+    let minAltitude = Infinity;
     /* eslint-disable-next-line */
     for (const id in coordinates.featureVertices) {
         // extract contour coodinates and properties of one feature
         const { contour, property } = extractFeature(coordinates, properties, id);
         // get altitude and extrude amount from properties
         const altitudeBottom = getAltitude(options, property);
+        minAltitude = Math.min(minAltitude, altitudeBottom);
```
Comment (peppsac) [inline src/Renderer/ThreeExtended/Feature2Mesh.js]: What is this `minAltitude` stuff? And why is that necessary?

File: examples/globe_wfs_extruded.js [hunk 12]
```diff
@@ -90,13 +92,37 @@ function acceptFeature(properties) {
...
+                mesh.scale.z = Math.min(
+                    1.0, mesh.scale.z + 0.016);
+                mesh.updateMatrixWorld(true);
+            }
+        }
+        meshes = meshes.filter(function filter(m) { return m.scale.z < 1; });
+    },
+};
+
+globeView.addFrameRequester(scaler);
```
Comment (peppsac) [inline examples/globe_wfs_extruded.js]: Nice to get this working on the globe!

**Review by peppsac — CHANGES_REQUESTED** [review]
Here's a first round of comments on the changes proposed.

File: index.html (hunk 1 above)
Comment (gchoqueux) [reply index.html]: ok

File: src/Core/Geographic/Coordinates.js (hunk 2 above)
Comment (gchoqueux) [reply src/Core/Geographic/Coordinates.js]: done

File: src/Core/Math/Ellipsoid.js (hunk 3 above)
Comment (gchoqueux) [reply src/Core/Math/Ellipsoid.js]: done

File: src/Core/Prefab/GlobeView.js (hunk 4 above)
Comment (gchoqueux) [reply src/Core/Prefab/GlobeView.js]: * `getCommonGeometryExtent` moved in tile builder
* `sizeLatitude` is modified
* `getCommonGeometryExtent`, in function of builder, it doesn't do modulo

File: src/Core/Prefab/GlobeView.js (hunk 5 above)
Comment (gchoqueux) [reply src/Core/Prefab/GlobeView.js]: changed in `getQuaternionFromExtent`

File: src/Core/Prefab/Planar/PlanarTileBuilder.js (hunk 6 above)
Comment (gchoqueux) [reply src/Core/Prefab/Planar/PlanarTileBuilder.js]: cleaned

File: src/Core/Scheduler/Providers/TileProvider.js (hunk 7 above)
Comment (gchoqueux) [reply src/Core/Scheduler/Providers/TileProvider.js]: yes, in another PR, there's not user cases for the moment

File: src/Core/Scheduler/Providers/TileProvider.js (hunk 8 above)
Comment (gchoqueux) [reply src/Core/Scheduler/Providers/TileProvider.js]: As the commands are asynchronous many tiles may ask for geometry during construction

File: src/Core/TileMesh.js (hunk 9 above)
Comment (gchoqueux) [reply src/Core/TileMesh.js:29]: obliged to clone because the DEM is different for each tile,

File: src/Renderer/ThreeExtended/Feature2Mesh.js (hunk 11 above)
Comment (gchoqueux) [reply src/Renderer/ThreeExtended/Feature2Mesh.js]: It's the minimum altitude of the vertex of the points. It's used to position the object3D on the surface normal

File: src/Core/Scheduler/Providers/TileProvider.js (hunk 8 above)
Comment (autra) [reply src/Core/Scheduler/Providers/TileProvider.js]: Everything here is synchronous, so it is guaranteed that no other part of itowns (or any other js in this page really) will get executed between this line and line 83. So the deferred will actually never be used here and you can directly set the TileGeometry in the cacheGeometry, without all the promise boilerplate.

Comment (gchoqueux) [reply src/Core/Scheduler/Providers/TileProvider.js]: However the commands are executed in asynchronous, many tiles can ask for geometry during construction.
I tested it didn't work

Comment (autra) [reply src/Core/Scheduler/Providers/TileProvider.js]: Please be more specific. What does "during construction" mean?

>  I tested it didn't work

What did you test? What didn't work?

> However the commands are executed in asynchronous

The fact that commands execute asynchronously has nothing to do with the matter. You set a promise on line 76 and resolve it on line 83.  No code can execute between these 2 lines, whether synchronous or asynchronous. 

Have you removed the `then` logic on line 87 when you tested?

Comment (gchoqueux) [reply src/Core/Scheduler/Providers/TileProvider.js]: Thanks @autra , I started this PR a long time ago, where commands was concurrent.
I remove the promise in provider

File: src/Core/Math/Ellipsoid.js [hunk 13]
```diff
@@ -56,10 +56,10 @@ Ellipsoid.prototype.setSize = function setSize(size) {
...
     // var n;
     var k = new THREE.Vector3();
-    var n = this.geodeticSurfaceNormalCartographic(coordCarto);
+    const n = normal ? normal.clone() : this.geodeticSurfaceNormalCartographic(coordCarto);
```
Comment (peppsac) [inline src/Core/Math/Ellipsoid.js]: You could replace this with:
```js
const n = normal ? normal.clone() : coordCarto.geodesicNormal.clone();
```
Or even better, you could avoid changing the API and you'll get the same result by doing:
```js
const n = coordCarto.geodesicNormal.clone();
```

File: src/Core/Prefab/Globe/BuilderEllipsoidTile.js [hunk 14]
```diff
@@ -79,9 +87,37 @@ BuilderEllipsoidTile.prototype.getUV_PM = function getUV_PM(params) {
...
+// return common extent to pool the geometries
+// the geometry in common extent is identical to the existing input
+// with a transformation (translation, rotation)
+BuilderEllipsoidTile.prototype.getCommonGeometryExtent = function getCommonGeometryExtentFn(extent) {
+    // Common geometry is looking for only on longitude
+    // TODO: It should be possible to use equatorial plan symetrie,
+    // but we should be reverse UV on tile
+    const sizeLongitude = Math.abs(extent.west() - extent.east()) / 2;
+    const communExtent = new Extent(extent.crs(), -sizeLongitude, sizeLongitude, extent.south(), extent.north());
+    communExtent._internalStorageUnit = extent._internalStorageUnit;
```
Comment (peppsac) [inline src/Core/Prefab/Globe/BuilderEllipsoidTile.js]: `commun` is not an English word :)

File: src/Core/Prefab/Globe/BuilderEllipsoidTile.js [hunk 15]
```diff
@@ -79,9 +87,37 @@ BuilderEllipsoidTile.prototype.getUV_PM = function getUV_PM(params) {
     return t - params.deltaUV1;
 };
 
+// return common extent to pool the geometries
+// the geometry in common extent is identical to the existing input
+// with a transformation (translation, rotation)
+BuilderEllipsoidTile.prototype.getCommonGeometryExtent = function getCommonGeometryExtentFn(extent) {
```
Comment (peppsac) [inline src/Core/Prefab/Globe/BuilderEllipsoidTile.js]: I'd prefer to reserve `getXXX` for simple getter method.

I have more comment about this API, see layer.

File: src/Core/Scheduler/Providers/TileProvider.js [hunk 16]
```diff
@@ -49,47 +42,78 @@ TileProvider.prototype.preprocessDataLayer = function preprocessLayer(layer, vie
...
+    const builder = layer.builder;
+    const parent = command.requester;
+    const level = (command.level === undefined) ? (parent.level + 1) : command.level;
 
-    var parent = command.requester;
+    if (!this.cacheGeometry[level]) {
+        this.cacheGeometry[level] = new Map();
+    }
 
+    const ce = builder.getCommonGeometryExtent(extent);
```
Comment (peppsac) [inline src/Core/Scheduler/Providers/TileProvider.js]: I don't think TileProvider needs to be aware of this: this could an internal feature of each tile builder.

But even without this, the API  could be simplified a lot if `getCommonGeometryExtent` was something that would return 2 things: the new extent and the transformation to apply.
And this would give you a better naming as well: something like: `computeSharableExtent`.

WDYT?

File: src/Core/Scheduler/Providers/TileProvider.js [hunk 17]
```diff
@@ -49,47 +42,78 @@ TileProvider.prototype.preprocessDataLayer = function preprocessLayer(layer, vie
...
+    const south = ce.south().toFixed(8);
+
+    let geometry = this.cacheGeometry[level].get(south);
```
Comment (peppsac) [inline src/Core/Scheduler/Providers/TileProvider.js]: I already commented about this: you can't use `south` as a cache key - and no it won't wait for another PR because it can break today.
Additional things to used for the key: builder type, segment count and probably disableSkirt.

Also I'd prefer a single cache than 1 cache per level.

Last but not least: when are you cleaning this cache?
If I zoom in a lot and zoom out, the geometries for high levels will never get disposed..

**Review by peppsac — CHANGES_REQUESTED** [review]
Looks better, thanks.

You didn't change the cleanup code (see ObjectRemovalHelper.removeChildrenAndCleanup) so geometries are going to be disposed even if they are still in used...

Another question: this change is introducing a behavior change - not an API one but a behavior one (tiles' transformation is different now). So you need to document this somewhere in your commit message with a BREAKING CHANGE tag so that `Release Notes` can grab this change.

File: src/Core/Prefab/Globe/BuilderEllipsoidTile.js [hunk 18]
```diff
@@ -33,24 +38,30 @@ BuilderEllipsoidTile.prototype.Prepare = function Prepare(params) {
...
+
     // let's avoid building too much temp objects
     params.projected = { longitudeRad: 0, latitudeRad: 0 };
 };
 
 // get center tile in cartesian 3D
-BuilderEllipsoidTile.prototype.Center = function Center(params) {
-    params.center = params.extent.center(this.tmp.coords[0])
+BuilderEllipsoidTile.prototype.Center = function Center(extent) {
+    this.tmp.coords[0]._normal = undefined;
```
Comment (peppsac) [inline src/Core/Prefab/Globe/BuilderEllipsoidTile.js]: Can you please move this to `Extent` (probably in Extent.center() method) or in `Coordinates` (in the Coordinates.set())?

File: src/Core/Prefab/Panorama/PanoramaTileBuilder.js [hunk 19]
```diff
@@ -97,57 +102,55 @@ PanoramaTileBuilder.prototype.vProjecte = function vProjecte(v, params) {
...
-        const max = new THREE.Vector3(length * 0.5, height * 0.5, thickness * 0.5);
-
-        const translate = new THREE.Vector3(0, 0, thickness * -0.5);
-        return new OBB(min, max, direction, translate);
-    }
+PanoramaTileBuilder.prototype.OBB = function _OBB(boundingBox) {
+    return new OBB(boundingBox.min, boundingBox.max);
+};
+
+PanoramaTileBuilder.prototype.getCommonGeometryExtent = function getCommonGeometryExtentFn(extent) {
```
Comment (peppsac) [inline src/Core/Prefab/Panorama/PanoramaTileBuilder.js]: Is this still used?

File: src/Core/Prefab/Panorama/PanoramaTileBuilder.js [hunk 20]
```diff
@@ -97,57 +102,55 @@ PanoramaTileBuilder.prototype.vProjecte = function vProjecte(v, params) {
...
+    // with a transformation (translation, rotation)
+
+    // TODO: It should be possible to use equatorial plan symetrie,
+    // but we should be reverse UV on tile
+    // Common geometry is looking for only on longitude
+    const sizeLongitude = Math.abs(extent.west() - extent.east()) / 2;
+    const sharableExtent = new Extent(extent.crs(), -sizeLongitude, sizeLongitude, extent.south(), extent.north());
+    sharableExtent._internalStorageUnit = extent._internalStorageUnit;
+
+    // compute rotation to transform tile to position it on ellipsoid
```
Comment (peppsac) [inline src/Core/Prefab/Panorama/PanoramaTileBuilder.js]: The comment refers to an "ellipsoid" but there's no ellipsoid in this context.
Can you please fix the comment?

File: src/Core/Scheduler/Providers/TileProvider.js [hunk 21]
```diff
@@ -49,47 +42,73 @@ TileProvider.prototype.preprocessDataLayer = function preprocessLayer(layer, vie
...
+
+    if (!this.cacheGeometry[level]) {
+        this.cacheGeometry[level] = new Map();
+    }
+
+    const { sharableExtent, quaternion, position } = builder.computeSharableExtent(extent);
+    const south = sharableExtent.south().toFixed(8);
 
-    var parent = command.requester;
+    console.log(quaternion, position);
```
Comment (peppsac) [inline src/Core/Scheduler/Providers/TileProvider.js]: This should probably be removed :)

File: src/Core/Scheduler/Providers/TileProvider.js [hunk 22]
```diff
@@ -49,47 +42,73 @@ TileProvider.prototype.preprocessDataLayer = function preprocessLayer(layer, vie
...
+    if (!geometry) {
+        const paramsGeometry = {
+            extent: sharableExtent,
+            level,
+            segment: layer.segments || 16,
+            disableSkirt: layer.disableSkirt,
+        };
+
+        geometry = new TileGeometry(paramsGeometry, builder);
+        this.cacheGeometry[level].set(south, geometry);
```
Comment (peppsac) [inline src/Core/Scheduler/Providers/TileProvider.js]: `level` isn't a good cache key (see earlier comment https://github.com/iTowns/itowns/pull/586#discussion_r157729520)

File: src/Core/Scheduler/Providers/TileProvider.js [hunk 23]
```diff
@@ -34,6 +26,7 @@ TileProvider.prototype.preprocessDataLayer = function preprocessLayer(layer, vie
 
     layer.level0Nodes = [];
     layer.onTileCreated = layer.onTileCreated || (() => {});
+    layer.object3d.invQuaternion = layer.object3d.quaternion.clone().inverse();
```
Comment (peppsac) [inline src/Core/Scheduler/Providers/TileProvider.js]: `invQuaternion` is never updated. And IMHO it's a bad idea to introduce this concept: just call inverse() when you need it.
Or maybe rewrite the code doing this to avoid requiring a matrix inversion (in worldToLocal) + 2 quaternion.inverse() call.

**Comment by gchoqueux** [comment]
@peppsac updated

File: src/Process/ObjectRemovalHelper.js [hunk 24]
```diff
@@ -8,7 +8,12 @@ export default {
             obj.dispose();
         } else {
             if (obj.geometry) {
-                obj.geometry.dispose();
+                if (obj.geometry.count) {
```
Comment (peppsac) [inline src/Process/ObjectRemovalHelper.js]: This isn't a good idea since you're breaking geometry disposal for other types of geometries.

See other comment  in TileProvider.js

File: src/Process/PanoramaTileProcessing.js [hunk 25]
```diff
@@ -1,3 +1,5 @@
+// import * as THREE from 'three';
```
Comment (peppsac) [inline src/Process/PanoramaTileProcessing.js]: Remove dead code please.

File: src/Core/Scheduler/Providers/TileProvider.js [hunk 26]
```diff
@@ -49,47 +41,80 @@ TileProvider.prototype.preprocessDataLayer = function preprocessLayer(layer, vie
...
+        const paramsGeometry = {
+            extent: sharableExtent,
+            level,
+            segment,
+            disableSkirt: layer.disableSkirt,
+        };
+
+        geometry = new TileGeometry(paramsGeometry, builder);
+        geometry._count = 0;
+        this.cacheGeometry.set(key, geometry);
```
Comment (peppsac) [inline src/Core/Scheduler/Providers/TileProvider.js]: You could do this:
```js
geometry.dispose = () => {
    geometry._count--;
    if (geometry._count == 0) {
       THREE.BufferGeometry.prototype.dispose.call(geometry);
       this.cacheGeometry.delete(key);
    }
}
```
And remove the `count` property. This way the ref-counting feature would be internal to TileProvider.

**Review by peppsac — COMMENTED** [review]
Did you observe some measurable memory or perf gains with this changeset?

**Comment by gchoqueux** [comment]
> Did you observe some measurable memory or perf gains with this changeset?

the processes time (to build tiles) and used memory (geometryTile) are divided by 3

File: src/Core/Scheduler/Providers/TileProvider.js [hunk 27]
```diff
@@ -49,47 +41,79 @@ TileProvider.prototype.preprocessDataLayer = function preprocessLayer(layer, vie
...
+        level,
+        materialOptions: layer.materialOptions,
     };
 
-    const geometry = new TileGeometry(params, command.layer.builder);
-
-    var tile = new TileMesh(geometry, params);
-
-    tile.layer = command.layer.id;
+    geometry.count++;
```
Comment (peppsac) [inline src/Core/Scheduler/Providers/TileProvider.js]: Shouldn't it be `geometry._count++`?

File: src/Core/Geographic/Extent.js [hunk 28]
```diff
@@ -245,6 +245,7 @@ Extent.prototype.center = function center(target) {
     let c;
     if (target) {
         Coordinates.call(target, this._crs, this._values[0], this._values[2]);
+        target._normal = undefined;
```
Comment (peppsac) [inline src/Core/Geographic/Extent.js]: Resetting the `_normal` value should be done inside `Coordinates`.

File: src/Core/Prefab/Globe/BuilderEllipsoidTile.js [hunk 29]
```diff
@@ -33,24 +40,27 @@ BuilderEllipsoidTile.prototype.Prepare = function Prepare(params) {
...
-    params.center = params.extent.center(this.tmp.coords[0])
+BuilderEllipsoidTile.prototype.Center = function Center(extent) {
+    return extent.center(this.tmp.coords[0])
         .as('EPSG:4978', this.tmp.coords[1]).xyz();
-    return params.center;
 };
 
 // get position 3D cartesian
 BuilderEllipsoidTile.prototype.VertexPosition = function VertexPosition(params) {
+    this.tmp.coords[0]._normal = undefined;
```
Comment (peppsac) [inline src/Core/Prefab/Globe/BuilderEllipsoidTile.js]: Proposal: move this reset in `Coordinates.set()` and use it here:
```js
this.tmp.coords[0].set(
    'EPSG:4326',
    params.projected.longitudeRad,
    params.projected.latitudeRad);
```

**Review by peppsac — COMMENTED** [review]
> the processes time (to build tiles) and used memory (geometryTile) are divided by 3

This info should be in one of the commits message - and maybe add explain you measured it.

The code is almost good, I just left a few comments to address.

**Comment by gchoqueux** [comment]
> the processes time (to build tiles) and used memory (geometryTile) are divided by 3

I compare master and this PR with different calls from the `GlobeControls` API :
 * Total running time on `TileProvider.executeCommand`
  * Snapshots of memory occupied by all instances of  `TileGeometry`

File: src/Core/Prefab/Globe/BuilderEllipsoidTile.js [hunk 30]
```diff
@@ -33,24 +40,30 @@ BuilderEllipsoidTile.prototype.Prepare = function Prepare(params) {
...
+    this.tmp.coords[0].set(
+        'EPSG:4326',
+        params.projected.longitudeRad,
+        params.projected.latitudeRad);
+
+    this.tmp.coords[0]._internalStorageUnit = UNIT.RADIAN;
 
     this.tmp.coords[0].as('EPSG:4978', this.tmp.coords[1]).xyz(this.tmp.position);
-    this.tmp.normal.copy(this.tmp.position).normalize();
+    this.tmp.normal.copy(this.tmp.coords[1].geodesicNormal);
```
Comment (peppsac) [inline src/Core/Prefab/Globe/BuilderEllipsoidTile.js]: Btw, you could remove `this.tmp.normal` completely and replace VertexNormal code by `return this.tmp.coords[1].geodesicNormal`.

**Review by peppsac — COMMENTED** [review]
Thanks for all the edits, the PR looks good now.

I left a small suggestion to simplify the code. And of course you need to rebase / edit commits properly before merging :)

**Comment by gchoqueux** [comment]
@peppsac  thanks for review
commit message updated

**Comment by gchoqueux** [comment]
@peppsac this PR can be merged

**Review by peppsac — APPROVED** [review]
Ok this PR looks good now; thanks for the hard work!

**Comment by peppsac** [comment]
@gchoqueux This PR broke the http://www.itowns-project.org/itowns/examples/multiglobe.html example.

Could you take a look at this please? (I can open a new issue if you prefer)
